nservicebus/handlers/accessing-data.md

```diff
@@ -2,70 +2,87 @@
+### Outbox
 
- * [SQL Persistence](/persistence/sql/accessing-data.md) (with SQL Server, MySQL, PostgreSQL, or Oracle)
- * [NHibernate](/persistence/nhibernate/accessing-data.md) (with SQL Server or Oracle)
- * [RavenDB](/persistence/ravendb/#shared-session)
+[The Outbox](/nservicebus/outbox) is a pattern that provides *exactly-once* message processing experience even when dealing with transport and database that don't support distributed transactions e.g. RabbitMQ and MongoDB. It achieves it by storing the incoming message ID and the outgoing messages in the same transaction as the business state change.
 
+The Outbox can be used with any transport and with any persistence component that [support the Synchronized Storage Session](#synchronized-storage-session).
 
-### Transport in native transaction mode
+Instead of preventing the duplicates, the Outbox detects them and ensures that the effects of processing duplicate messages are ignored (not persisted).
 
-In this mode, the NServiceBus-managed data store context can be committed multiple times for a single message and it is up to the user to guarantee *idempotency*. The difference between user-managed connections, though, is the fact that data store context is shared between the handlers so there might be one (possibly generic) handler that takes care of the *idempotency*, allowing others to focus on pure business problems.
 
-snippet: BusinessData-Native-Managed
+## Manual de-duplication
 
+In situations where neither of the built-in de-duplication strategies can be applied, the de-duplication of messages needs to be handled at the application level, in the message handler itself. In these cases the [Synchronized Storage Session](#synchronized-storage-session) should not be used and each handler should guarantee the idempotence of its behavior.
 
-### Outbox
+### Idempotence caveats
+
+Message processing logic is idempotent if it can be applied multiple times and the outcome is the same as if it was applied once. The outcome includes both the application state changes and the potential outgoing messages sent. Consider the following pseudocode:
+
+snippet: BusinessData-ManualIdempotence
+
+and think about the behavior of the message processing:
+
+ - NServiceBus by default defers sending out messages until message handler has completed so the behavior of the code above is as if the call to `Send` **was after** the call to `ModifyState`
+ - If outgoing messages are sent before state change is committed (e.g if code above used [immediate dispatch](/nservicebus/messaging/send-a-message.md#dispatching-a-message-immediately)) there is a risk of creating *ghost messages* -- messages that carry the state change that has never been made durable
+ - If outgoing messages are sent after the state change is committed there is risk of message loss if the send operation fails. To prevent this, the outgoing messages need to be sent **even if it appears to be a duplicate**
```

Comment:
Should you also talk about the case in which business state changes succeed, outgoing messages are sent successfully, BUT the message in which the current handler is being executed for is failed to be reported back to the transport that processing has succeeded which in turn would cause a retry/redelivery?